File: 9/9a.py
with open("9.txt") as file:
    lines = file.read().split('\n')

    visitedSpaces = {'0,0': True}

    class Coordinates:
        def __init__(self, x, y):
            self.x = x
            self.y = y

    head = Coordinates(0, 0)
    tail = Coordinates(0, 0)

    for line in lines:
        (direction, count) = line.split(' ')
        for i in range(int(count)):
            if direction == 'L':
                head.x -= 1
            elif direction == 'R':
                head.x += 1
            elif direction == 'U':
                head.y -= 1
            elif direction == 'D':
                head.y += 1

            xDiff = head.x - tail.x
            yDiff = head.y - tail.y

            if abs(xDiff) == 2:
                tail.x += int(xDiff/2)
                tail.y += yDiff
            elif abs(yDiff) == 2:
                tail.y += int(yDiff/2)
                tail.x += xDiff

            visitedSpaces[f'{tail.y},{tail.x}'] = True

    print(len(visitedSpaces))

File: 9/9a.js
const fs = require('fs');

const input = fs.readFileSync('./9.txt', 'utf8');

const lines = input.split(/\r?\n/);

const visitedSpaces = { '0,0': true };

const head = { x: 0, y: 0 };
const tail = { x: 0, y: 0 };

for (const line of lines) {
  const [dir, count] = line.split(' ');
  for (let i = 0; i < count; i++) {
    if (dir === 'L') head.x -= 1;
    else if (dir === 'R') head.x += 1;
    else if (dir === 'U') head.y -= 1;
    else if (dir === 'D') head.y += 1;

    const xDiff = head.x - tail.x;
    const yDiff = head.y - tail.y;

    // pair of diffs that cause moves can be:
    // abs(2) in one direction and abs(0) or abs(1) in the other
    if (Math.abs(xDiff) === 2) {
      tail.x += xDiff / 2;
      tail.y += yDiff;
    } else if (Math.abs(yDiff) === 2) {
      tail.y += yDiff / 2;
      tail.x += xDiff;
    }

    visitedSpaces[tail.y + ',' + tail.x] = true;
  }
}

console.log(Object.keys(visitedSpaces).length);
